File: cpp/src/_438.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<int> findAnagrams(string s, string p)
    {
        int left, right;
        int len = s.size();
        int lenp = p.size();
        unordered_map<char,int> hashmap;
        unordered_map<char,int> ha;
        vector<int> ans;
        for (const char &c : p)
        {
            if(hashmap.find(c) == hashmap.end())
            {
                hashmap[c] = 1;
            }
            else
            {
                ++hashmap[c];
            }
            ha[c] = 0;
        }
        int cnt = 0;
        for (left = right = 0; left < len - lenp + 1; ++left)
        {
            if (hashmap.count(s[left]) == 0)
            {
                right = left + 1;
                continue;
            }
            while (right < len && hashmap.count(s[right]) > 0 && ha[s[right]] < hashmap[s[right]])
            {
                ++cnt;
                ++ha[s[right]];
                ++right;
                if (cnt == lenp)
                {
                    ans.emplace_back(left);
                    break;
                }
            }
            --cnt;
            --ha[s[left]]; // left pointer move
        }
        return ans;
    }
};

int main()
{
    Solution test;
    auto a = test.findAnagrams("baa", "aa");
    for (auto i : a)
    {
        cout << i << endl;
    }
    return 0;
}

File: 438.py
from typing import List
from collections import Counter

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        cnt_w = Counter()
        cnt_p = Counter(p)
        ans = []
        for i, x in enumerate(s):
            cnt_w[x] += 1
            left = i - len(p) + 1
            if left < 0:
                continue
            if cnt_p == cnt_w:
                ans.append(left)
            out = s[left]
            cnt_w[out] -= 1
            if cnt_w[out] == 0:
                del cnt_w[out]
        return ans
    def findAnagrams(self, s: str, p: str) -> List[int]:
        cnt = Counter(p)
        ans = []
        left = 0
        for right, x in enumerate(s):
            cnt[x] -= 1
            while cnt[x] < 0:
                cnt[s[left]] += 1
                left += 1
            if right == left + len(p) - 1:
                ans.append(left)
        return ans

t = Solution()
t.findAnagrams("cbaebabacd", "abc")